tests/pipeline/test_engine.py

```diff
@@ -372,52 +381,59 @@ def test_masked_factor(self):
-            # another factor to the pipeline term graph does not cause any
-            # unexpected exceptions when calling `run_pipeline`.
-            for factor2 in (None,
-                            RollingSumDifference(mask=NotSpecified),
-                            RollingSumDifference(mask=mask)):
-                if factor2 is None:
-                    columns = {'factor1': factor1}
-                else:
-                    columns = {'factor1': factor1, 'factor2': factor2}
-                pipeline = Pipeline(columns=columns)
-                results = engine.run_pipeline(pipeline, dates[0], dates[-1])
-                factor1_results = results['factor1'].unstack()
-
-                expected = {
-                    asset: values if asset.sid <= asset_id else missing_values
-                    for asset in assets
-                }
-
-                assert_frame_equal(
-                    factor1_results,
-                    DataFrame(expected, index=dates, columns=assets),
-                )
+        missing_values = full(num_dates, nan)
+        factor1_values = full(num_dates, constants[open], dtype=float64)
+        factor2_values = full(num_dates,
+                              3.0 * (constants[open] - constants[close]))
+
+        def create_expected_results(values):
+            expected_values = {}
+            for asset_id, asset in zip(asset_ids, assets):
+                # Subtract 1 because asset_id is 1-indexed.
+                num_values = num_dates - (asset_id - 1)
+                expected_values[asset] = \
+                    append(values[:num_values], missing_values[num_values:])
+            return DataFrame(expected_values, index=dates, columns=assets)
+
+        factor1_expected = create_expected_results(factor1_values)
+        factor2_expected = create_expected_results(factor2_values)
+
+        # Produce a mask that looks like:
+        #
+        #       Equity(0 [A])  Equity(1 [B])  Equity(2 [C])  Equity(3 [D])
+        # Day 1          True           True           True          False
+        # Day 2          True           True          False          False
+        # Day 3          True          False          False          False
+        #
+        mask = AssetIDPlusDay() < asset_ids[-1] + dates[0].day
+
+        # Test running a pipeline with a single masked factor.
+        columns = {'factor1': OpenPrice(mask=mask)}
+        pipeline = Pipeline(columns=columns)
+        results = engine.run_pipeline(pipeline, dates[0], dates[-1])
+        factor1_results = results['factor1'].unstack()
+        assert_frame_equal(factor1_results, factor1_expected)
+
+        # Test running a pipeline with a second factor. This ensures that
+        # adding another factor to the pipeline with a different window length
+        # does not cause any unexpected behavior, especially when both factors
+        # share the same mask.
+        columns['factor2'] = RollingSumDifference(mask=mask)
```

Comment:
Can we also test that adding the second factor doesn't affect the result of the first factor?